R/type-tibble.R

```diff
@@ -32,3 +32,32 @@ vec_ptype2.data.frame.tbl_df <- function(x, y, ..., x_arg = "x", y_arg = "y") {
     y_arg = y_arg
   )
 }
+
+
+# Conditionally registered in .onLoad()
+vec_cast.tbl_df <- function(x, to, ..., x_arg = "x", to_arg = "to") {
+  UseMethod("vec_cast.tbl_df")
+}
+vec_cast.tbl_df.default <- function(x, to, ..., x_arg = "x", to_arg = "to") {
+  vec_default_cast(x, to, x_arg = x_arg, to_arg = to_arg)
+}
+vec_cast.tbl_df.tbl_df <- function(x, to, ..., x_arg = "x", to_arg = "to") {
+  tib_cast(x, to, x_arg = x_arg, to_arg = to_arg)
+}
+
+vec_cast.data.frame.tbl_df <- function(x, to, ..., x_arg = "x", to_arg = "to") {
+  tib_cast(x, to, x_arg = x_arg, to_arg = to_arg)
+}
+vec_cast.tbl_df.data.frame <- function(x, to, ..., x_arg = "x", to_arg = "to") {
+  df_cast(x, to, x_arg = x_arg, to_arg = to_arg)
+}
```

Comment:
Doesn't this still allow inheritance? It needs the checks like we have in glue about being a "bare tibble" or "bare data frame" (or we need a better system for that).

``` r
library(vctrs)
library(tibble)
devtools::load_all("~/Desktop/r/packages/vctrs")
#> Loading vctrs

x <- subclass(tibble())

vec_cast(x, x)
#> # A tibble: 0 x 0

class(vec_cast(x, x))
#> [1] "tbl_df"     "tbl"        "data.frame"
```

<sup>Created on 2020-03-17 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Reply:
Yes but these aren't called explicitly in the bare tibble / df cases since they are now handled at C level. I didn't guard for inheritance at R level because I think that instead of making the implementation of methods more complex, we should fix the dispatch mechanism.